src/OrchardCore.Modules/OrchardCore.Security/Services/SecurityHeadersMiddleware.cs

```diff
@@ -0,0 +1,29 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using OrchardCore.Security.Options;
+
+namespace OrchardCore.Security.Services
+{
+    public class SecurityHeadersMiddleware
+    {
+        private readonly SecurityHeadersOptions _options;
+        private readonly RequestDelegate _next;
+
+        public SecurityHeadersMiddleware(IOptions<SecurityHeadersOptions> options, RequestDelegate next)
+        {
+            _options = options.Value;
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            foreach (var provider in _options.HeaderPolicyProviders)
+            {
+                provider.ApplyPolicy(context);
```

Comment:
So here, as mentioned, some providers would need to have done some pre-processing.